### ISSUE TITLE
Smart contract wallets using predicates and NFTs (state minimized approach)

### ISSUE THREAD

**Issue by SilentCicero** [FuelLabs/fuel-specs#527]
# Abstract
Many developers would like to use native level predicates as a smart contract wallet. This usually presents a problem that predicates are stateless and thus have no way to manage state.

## Solution
Use a single native level NFT to represent state of the predicate wallet.

This means, a single smart contract wallet would only be using a single NFT UTXO as their entire state representation, which is vastly less than its smart contract counterparts.

## Pre-Construction:
- Unique NFT is minted with sub asset ID X
- Build the hash for predicate wallet W which looks for sub asset ID X as state
- Build the hash for state storing predicate Y which looks for W as the owner to spend to be present in the transaction and contains state hash S, ideally Y has the logic to unwrap W and ensure only W met spending conditions are fulfilled in the transaction
- NFT X is sent to state storing predicate Y

## Transaction Construction
- Predicate wallet W looks for sub asset ID X in the current transaction T
- W unwraps the code at found predicate Y for asset X and reveals the state hash S
- W looks for the preimage of S in the transaction T found in witness W
- When updating state, W creates a new state predicate Y2 containing the new state hash for S2 with enforcement still remaining on W
- W sends state NFT X to the new state predicate at Y2

**Witnesses:**
- Pre-image state for S
- Signatures for owner W

**Inputs:**
- Asset A with owner W
- NFT X with owner Y

**Outputs:**
- Asset A sent to desired owner
- NFT X with new owner Y2

TLDR: we are using the NFT identifier as a constant identifier for state and the owner as the recorded state root. State and pre-images have to be re-exposed every time you spend over bandwidth, but bandwidth could be reduced further by using LDC, and likely one other read-only contract input.
